src/runtime/distributed/worker/MPIHelper.h

```diff
@@ -0,0 +1,201 @@
+
+
+#include <vector>
+
+#define COORDINATOR 0
+
+enum TypesOfMessages{
+    BROADCAST=0, DATASIZE, DATA, DATAACK, MLIRSIZE, MLIR, INPUTKEYS, OUTPUT, OUTPUTKEY,  DETACH, OBJECTIDENTIFIERSIZE, OBJECTIDENTIFIER
+};
+enum WorkerStatus{
+    LISTENING=0, DETACHED, TERMINATED
+};
+
+
+class MPIHelper{
+
+        public:
+        static int getCommSize(){
+            int worldSize;
+            MPI_Comm_size(MPI_COMM_WORLD, &worldSize);
+            return worldSize;    
+        }
+        
+        static WorkerImpl::StoredInfo constructStoredInfo(std::string input)
+        {
+            WorkerImpl::StoredInfo info;
+            std::stringstream s_stream(input);
+            std::vector<std::string> results;
+            while(s_stream.good()) {
+                std::string substr;
+                getline(s_stream, substr, ','); //get first string delimited by comma
+                results.push_back(substr);
+            }
+            info.identifier=results.at(0);
+            sscanf(results.at(1).c_str() , "%zu", &info.numRows);
+            sscanf(results.at(2).c_str() , "%zu", &info.numCols);
+            return info; 
+        }
+       
+        static distributed::Data getResults(int *rank){
+            size_t resultsLen=0;
+            void * results;
+            distributed::Data matProto;
+            getMessage(rank, OUTPUT, MPI_UNSIGNED_CHAR ,&results, &resultsLen);
+            MPISerializer::deserializeStructure(&matProto, results , resultsLen);
+           // std::cout<<"got results from "<<*rank<<std::endl;     
+            free(results);
+            return matProto;
+        }
+       
+        static WorkerImpl::StoredInfo getDataAcknowledgement(int *rank){
+            char * dataAcknowledgement;
+            size_t len;
+            getMessage(rank, DATAACK, MPI_CHAR, (void **)&dataAcknowledgement, &len);
+            std::string incomeAck = std::string(dataAcknowledgement);
+            WorkerImpl::StoredInfo info=constructStoredInfo(incomeAck);
+            free(dataAcknowledgement);
+            return info;  
+        }
+        static void sendObjectIdentifier(std::string identifier, int rank)
+        {
+            int len = identifier.length();
+            len++;
+            MPI_Send(&len,1, MPI_INT, rank, OBJECTIDENTIFIERSIZE, MPI_COMM_WORLD);
+            char message[len];
+            std::strcpy(message, identifier.c_str());
+            message[len-1]='\0'; 
+            MPI_Send(message,len, MPI_CHAR, rank, OBJECTIDENTIFIER, MPI_COMM_WORLD);
+
+        }
+        static void sendData(size_t messageLength, void * data){
+            int worldSize=getCommSize();
+            int  message= messageLength;
+            for(int rank=0; rank<worldSize;rank++)
+            {
+                if(rank==COORDINATOR)
+                    continue;          
+                MPI_Send(&message,1, MPI_INT, rank, BROADCAST, MPI_COMM_WORLD);                    
```

Comment:
If I understood your comment correctly, I would say I do not think it will be faster, i.e., multiple threads performing the calls require support from the MPI library, i.e.,  MPI_Init_thread. Furthermore, since the message size is one integer, it will be within the eager limit and it won't block the master anyway.
Last thing, we can completely remove the pattern of sending one message before the actual data message and rely only on the tags. However, we won't be able to use the MPU_Broadcast as it does not have a tag associated with it, and we will have to use MPI_recev at the worker and lose the potential performant tree implementation of MPI_Broadcast